File: packages/python/src/Seguranca/Ataques/AtaqueAproximacaoLinear.py
"""Aproximação linear (criptoanálise de Matsui): procura por correlação entre
bits de entrada (IV) e bits de saída do keystream. Uma cifra ideal não tem
aproximação linear com viés relevante; bias alto é uma pista explorável.
"""

from ..ResultadoAtaque import ResultadoAtaque
from ..Util import random_bytes


class AtaqueAproximacaoLinear:
    def __init__(self, amostras: int = 200, tamanho_bloco: int = 32, limite_bias: float = 0.3):
        self.amostras = amostras
        self.tamanho_bloco = tamanho_bloco
        self.limite_bias = limite_bias

    def nome(self) -> str:
        return "Aproximação linear (viés de Walsh)"

    def executar(self, alvo) -> ResultadoAtaque:
        tam_iv = alvo.tamanho_iv()
        bits_entrada = tam_iv * 8
        bits_saida = self.tamanho_bloco * 8

        entradas = []
        saidas = []

        for _ in range(self.amostras):
            chave = random_bytes(len(alvo.chave_de_teste().encode()))
            iv = random_bytes(tam_iv)
            ks = alvo.gerar_keystream_bruto(chave, iv, "enc", self.tamanho_bloco)

            in_bits = [(iv[j >> 3] >> (7 - (j & 7))) & 1 for j in range(bits_entrada)]
            out_bits = [(ks[j >> 3] >> (7 - (j & 7))) & 1 for j in range(bits_saida)]
            entradas.append(in_bits)
            saidas.append(out_bits)

        maior_bias = 0.0
        par = (-1, -1)
        for i in range(bits_entrada):
            for j in range(bits_saida):
                iguais = sum(1 for s in range(self.amostras) if entradas[s][i] == saidas[s][j])
                bias = abs(iguais / self.amostras - 0.5)
                if bias > maior_bias:
                    maior_bias = bias
                    par = (i, j)

        vulneravel = maior_bias > self.limite_bias

        return ResultadoAtaque(
            self.nome(),
            vulneravel,
            "alta" if vulneravel else "info",
            f"maior viés |p-0.5|={maior_bias:.4f} na aproximação "
            f"IV[bit {par[0]}] -> saída[bit {par[1]}] (limite {self.limite_bias}); "
            f"{self.amostras} amostras",
            {"maior_bias": maior_bias, "par": par},
        )

File: packages/typescript/src/Seguranca/Ataques/AtaqueAproximacaoLinear.ts
import { randomBytes } from '../Util.ts';
import { ResultadoAtaque } from '../ResultadoAtaque.ts';
import type { AtaqueInterface } from '../AtaqueInterface.ts';
import type { AlvoCriptografico } from '../AlvoCriptografico.ts';

/**
 * Aproximação linear (criptoanálise de Matsui): procura por correlação entre
 * bits de entrada (IV) e bits de saída do keystream. Uma cifra ideal não tem
 * aproximação linear com viés relevante; bias alto é uma pista explorável.
 */
export class AtaqueAproximacaoLinear implements AtaqueInterface {
  private readonly amostras: number;
  private readonly tamanhoBloco: number;
  private readonly limiteBias: number;

  constructor({
    amostras = 200,
    tamanhoBloco = 32,
    limiteBias = 0.3,
  }: { amostras?: number; tamanhoBloco?: number; limiteBias?: number } = {}) {
    this.amostras = amostras;
    this.tamanhoBloco = tamanhoBloco;
    this.limiteBias = limiteBias;
  }

  nome(): string {
    return 'Aproximação linear (viés de Walsh)';
  }

  executar(alvo: AlvoCriptografico): ResultadoAtaque {
    const tamIv = alvo.tamanhoIv();
    const bitsEntrada = tamIv * 8;
    const bitsSaida = this.tamanhoBloco * 8;

    const entradas: Uint8Array[] = [];
    const saidas: Uint8Array[] = [];

    for (let s = 0; s < this.amostras; s++) {
      const chave = randomBytes(Buffer.byteLength(alvo.chaveDeTeste(), 'utf8'));
      const iv = randomBytes(tamIv);
      const ks = alvo.gerarKeystreamBruto(chave, iv, 'enc', this.tamanhoBloco);

      const inBits = new Uint8Array(bitsEntrada);
      for (let j = 0; j < bitsEntrada; j++) {
        inBits[j] = (iv[j >> 3] >> (7 - (j & 7))) & 1;
      }
      const outBits = new Uint8Array(bitsSaida);
      for (let j = 0; j < bitsSaida; j++) {
        outBits[j] = (ks[j >> 3] >> (7 - (j & 7))) & 1;
      }
      entradas.push(inBits);
      saidas.push(outBits);
    }

    let maiorBias = 0;
    let par: [number, number] = [-1, -1];
    for (let i = 0; i < bitsEntrada; i++) {
      for (let j = 0; j < bitsSaida; j++) {
        let iguais = 0;
        for (let s = 0; s < this.amostras; s++) {
          if (entradas[s][i] === saidas[s][j]) iguais++;
        }
        const bias = Math.abs(iguais / this.amostras - 0.5);
        if (bias > maiorBias) {
          maiorBias = bias;
          par = [i, j];
        }
      }
    }

    const vulneravel = maiorBias > this.limiteBias;

    return new ResultadoAtaque(
      this.nome(),
      vulneravel,
      vulneravel ? 'alta' : 'info',
      `maior viés |p-0.5|=${maiorBias.toFixed(4)} na aproximação ` +
        `IV[bit ${par[0]}] -> saída[bit ${par[1]}] (limite ${this.limiteBias}); ` +
        `${this.amostras} amostras`,
      { maior_bias: maiorBias, par },
    );
  }
}
